pkg/providers/kube/fake.go

```diff
@@ -57,19 +63,70 @@ func (f fakeClient) ListEndpointsForIdentity(serviceIdentity identity.ServiceIde
 	panic(fmt.Sprintf("You are asking for K8sServiceAccount=%s but the fake Kubernetes client has not been initialized with this. What we have is: %+v", sa, f.svcAccountEndpoints))
 }
 
-func (f fakeClient) GetServicesForServiceAccount(svcAccount identity.K8sServiceAccount) ([]service.MeshService, error) {
-	services, ok := f.services[svcAccount]
+func (f fakeClient) GetServicesForServiceIdentity(serviceIdentity identity.ServiceIdentity) ([]service.MeshService, error) {
+	sa := serviceIdentity.ToK8sServiceAccount()
+	services, ok := f.services[sa]
 	if !ok {
-		return nil, errors.Errorf("ServiceAccount %s is not in cache: %+v", svcAccount, f.services)
+		return nil, errors.Errorf("ServiceAccount %s is not in cache: %+v", sa, f.services)
+	}
+	return services, nil
+}
+
+func (f fakeClient) ListServices() ([]service.MeshService, error) {
+	var services []service.MeshService
+
+	for _, svcs := range f.services {
+		services = append(services, svcs...)
 	}
 	return services, nil
 }
 
+func (f fakeClient) ListServiceIdentitiesForService(svc service.MeshService) ([]identity.ServiceIdentity, error) {
+	var serviceIdentities []identity.ServiceIdentity
+
+	for svcID := range f.services {
+		serviceIdentities = append(serviceIdentities, svcID.ToServiceIdentity())
+	}
+	return serviceIdentities, nil
+}
+
+func (f fakeClient) GetHostnamesForService(svc service.MeshService, locality service.Locality) ([]string, error) {
+	var hostnames []string
+
+	serviceName := svc.Name
+	namespace := svc.Namespace
+	port := tests.Endpoint.Port
+
+	fmt.Println(locality, service.LocalNS, "Whit")
+
+	if locality == service.LocalNS {
+		hostnames = append(hostnames, serviceName)
+	}
+
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s", serviceName, namespace))                   // service.namespace
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s.svc", serviceName, namespace))               // service.namespace.svc
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s.svc.cluster", serviceName, namespace))       // service.namespace.svc.cluster
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, namespace)) // service.namespace.svc.cluster.local
+	if locality == service.LocalNS {
+		// Within the same namespace, service name is resolvable to its address
+		hostnames = append(hostnames, fmt.Sprintf("%s:%d", serviceName, port)) // service:port
+	}
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s:%d", serviceName, namespace, port))                   // service.namespace:port
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s.svc:%d", serviceName, namespace, port))               // service.namespace.svc:port
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s.svc.cluster:%d", serviceName, namespace, port))       // service.namespace.svc.cluster:port
+	hostnames = append(hostnames, fmt.Sprintf("%s.%s.svc.cluster.local:%d", serviceName, namespace, port)) // service.namespace.svc.cluster.local:port
+	return hostnames, nil
+}
+
```

Comment:
Please avoid using fakes with the exact same logic as the actual function being tested. This is not maintainable and will introduce bugs quickly when the logic gets more complex. Instead, use mocks when managing fakes gets out of hand.

Would it be possible to just return hardcoded hostnames here or something the test can set when initializing the fake client's struct?

Reply:
the output is dependent on the combination of 2 inputs.. so you could have a double map.. but i'm not sure that is great.

I think this more so highlights the negatives of using fakes in the first place... maybe file a ticket to remove all usage of fakes and mocks altogehter?

at its heart it should probably just be faking the calls to the kubernetes master and nothing else. The k8s client simply has logic and doesn't make any network/sys calls anyways.

Reply:
anyways all that to say i don't think this is terrible to have logic twice if the decision has been made to "fake" logic.. why not get closer to to the real thing anyways?

Reply:
We very much need to use mocks and have moved away from using fakes as much as possible.  Just doesn't seem to be a good idea to duplicate actual code into fakes.

Reply:
can you provide a scenario that could lead to an issue? If the original changes, then the test will either break, or the potentially the tests weren't robust enough and they don't break (either way nothing detrimental has occurred). In fact it provides more flexibility than hardcoding values

And 2 cents on mocks: only network/syscalls should be mocked. It would offer much more robust testing if there was no mock provider or meshcatalog

Reply:
> can you provide a scenario that could lead to an issue? If the original changes, then the test will either break, or the potentially the tests weren't robust enough and they don't break (either way nothing detrimental has occurred). In fact it provides more flexibility than hardcoding values
> 

It is obvious that copying actual code into fakes is a maintenance hassle. When the actual code changes, if tests don't offer 100% coverage it is likely existing tests could pass while the fakes rely on an implementation that is no longer up to date. 

> And 2 cents on mocks: only network/syscalls should be mocked. It would offer much more robust testing if there was no mock provider or meshcatalog

This is something that I disagree with. If done right, mocking allows unit testing pieces of code independently of external components that the test doesn't care about. Inputs that depend on external components (could be a different interface, or whatever the test doesn't care about) can be easily mocked allowing the function to be well tested. There are benefits to using fakes for simple scenarios but it becomes unmanageable when a lot of code gets hardcoded into fakes (which I am not a fan of). If we want to leverage fakes, we should make them configurable, just like k8s fake clientset used in tests. We have files under `tests/scenarios/` that only use the fake clientset provided by k8s, but they are also harder to write and unit test because it involves initializing and testing the entire control plane components and not just a single function.

Reply:
Spoke  offline and resolved it